convergence/do_all_the_convergence_tests/compile_and_run.py

```diff
@@ -0,0 +1,250 @@
+                raise
+    os.chdir("build_convergence")
+    for prec in precision:
+        for o in orders:
+            compile_cmd = (
+                "CC={} CXX={} FC={} cmake .. "
+                "-DCMAKE_BUILD_TYPE=Release "
+                "-DCOMMTHREAD={} "
+                "-DEQUATIONS={} "
+                "-DGEMM_TOOL_LIST=LIBXSMM,PSpaMM, "
+                "-DHOST_ARCH={} "
+                "-DNUMBER_OF_MECHANISMS={} "
+                "-DORDER={} "
+                "-DPRECISION={} "
+                "-DTESTING=OFF && make -j8".format(
+                    compilers[0],
+                    compilers[1],
+                    compilers[2],
+                    on_off(on_a_cluster),
+                    equations,
+                    host_arch,
+                    num_mechs(equations),
+                    o,
+                    prec,
+                )
+            )
+            print(compile_cmd)
+            try:
+                subprocess.check_output(compile_cmd, shell=True)
+            except subprocess.CalledProcessError as cpe:
+                print("Build command exited with : {}".format(cpe.returncode))
+                print(cpe.output)
+                quit()
+            num_quantities = 9 + 6 * num_mechs(equations)
+            sn = seissol_name(arch_name(prec, host_arch), equations, o, "Release")
+            copy_source = sn
+            copy_target = os.path.join(cwd, sn)
+            copy(copy_source, copy_target)
+
+os.chdir(cwd)
+
+if args.steps in ["prepare", "all"]:
+    for n in resolutions:
+        for scale in scales:
+            generate_cmd = (
+                "cubeGenerator "
+                "-b 6 -x {0} -y {0} -z {0} "
+                "--px {1} --py {1} --pz {1} "
+                "-o {2}.nc -s {3}".format(2 ** n, parts[n], cube_name(n, scale), scale)
+            )
+            os.system(generate_cmd)
+        parameters = Path("parameters.template").read_text()
+        parameters = parameters.replace("MESH_FILE", cube_name(n, scale_map[equations]))
+        parameters = parameters.replace("END_TIME", end_time[equations])
+        parameters = parameters.replace("MATERIAL_FILE", material_file[equations])
+        parameters = parameters.replace(
+            "INITIAL_CONDITION", initial_condition[equations]
+        )
+        with open(par_name(equations, n), "w") as f:
+            f.write(parameters)
+
+if args.steps in ["run", "all"]:
+    for arch in archs:
+        for o in orders:
+            for n in resolutions:
+                log_file = os.path.join(log_dir, log_name(arch, equations, o, n))
+                nodes = parts[n] ** 3
+                if not on_a_cluster:
+                    run_cmd = "OMP_NUM_THREADS=8 ./{} {}".format(
+                        seissol_name(arch, equations, o), par_name(equations, n)
+                    )
+                    run_cmd += " > " + log_file
+                    print(run_cmd)
+                    os.system(run_cmd)
```

Comment:
Maybe a bit confusing that there are two different modes of execution and also two modes of settings (here OMP_NUM_THREADS=8 is hardcoded for on_a_cluster=false whereas on a cluster one would edit job.template).

Given that job scripts can also be run as normal bash scripts, you could also create a "job script" for local execution. I imagine that you could select a job template via the argument parser and then have one job template for local execution and X job templates for X clusters. Then the work-flow would be somewhat unified for local execution and cluster.

Reply:
Agreed, now there's a section in the code with some cluster specific settings and a job template.